File: hosts/android/app/src/main/java/dev/otagateway/host/HostStateStore.kt
package dev.otagateway.host

import android.content.Context
import android.util.Log
import com.callstack.reactnativebrownfield.OnMessageListener
import com.callstack.reactnativebrownfield.ReactNativeBrownfield
import org.json.JSONException
import org.json.JSONObject

/**
 * The native store behind the RN saved-state seam (apps/mobile
 * src/brownfield/host-state.ts).
 *
 * RN components checkpoint state slices by posting
 * `{ "type": "saveState", "key": <slice>, "state": <object> }` over the
 * brownfield message bridge; [register] writes each slice into
 * SharedPreferences. Every RN surface the host mounts gets the WHOLE store
 * back as the `savedStateJson` initial property ([readAllJson]), so a
 * dismissed component -- the fidget spinner mid-coast, say -- resumes exactly
 * where it left off, across tab switches, pushed screens, and process death.
 *
 * Slices are stored as JSON strings keyed by the RN-side slice key; malformed
 * slices are dropped on read (never propagated to a fresh surface).
 */
object HostStateStore {
    private const val TAG = "HostStateStore"
    private const val NAME = "dev.otagateway.host.state"
    private const val SAVE_STATE_MESSAGE_TYPE = "saveState"

    private fun prefs(context: Context) =
        context.getSharedPreferences(NAME, Context.MODE_PRIVATE)

    /** Subscribes to the brownfield bridge and persists saveState messages. */
    fun register(context: Context) {
        val appContext = context.applicationContext
        ReactNativeBrownfield.shared.addMessageListener(
            OnMessageListener { message ->
                parseSaveState(message)?.let { (key, stateJson) ->
                    prefs(appContext).edit().putString(key, stateJson).apply()
                }
            },
        )
    }

    /** The whole store as a JSON object of slices: `{"spinner": {...}, ...}`. */
    fun readAllJson(context: Context): String {
        val all = JSONObject()
        for ((key, value) in prefs(context).all) {
            if (value !is String) continue
            try {
                all.put(key, JSONObject(value))
            } catch (e: JSONException) {
                Log.w(TAG, "Dropping malformed saved-state slice '$key'", e)
            }
        }
        return all.toString()
    }

    /**
     * Extracts (key, state-JSON) from a raw bridge message, or null if it is
     * not a well-formed saveState message. The bridge is an untrusted input
     * source: anything malformed is ignored, matching BrownfieldReloadHandler.
     */
    private fun parseSaveState(message: String): Pair<String, String>? =
        try {
            val json = JSONObject(message)
            val key = json.optString("type").takeIf { it == SAVE_STATE_MESSAGE_TYPE }
                ?.let { json.optString("key").takeIf(String::isNotEmpty) }
            val state = json.optJSONObject("state")
            if (key != null && state != null) key to state.toString() else null
        } catch (e: JSONException) {
            Log.d(TAG, "Ignoring malformed brownfield message", e)
            null
        }
}

File: hosts/ios/OtaHost/HostStateStore.swift
import Foundation

/// The native store behind the RN saved-state seam (apps/mobile
/// src/brownfield/host-state.ts).
///
/// RN components checkpoint state slices by posting
/// `{ "type": "saveState", "key": <slice>, "state": <object> }` over the
/// brownfield message bridge; `BrownfieldBootstrap` routes those here and each
/// slice is written into `UserDefaults`. Every RN surface the host mounts gets
/// the WHOLE store back as the `savedStateJson` initial property
/// (`readAllJson`), so a dismissed component -- the fidget spinner mid-coast,
/// say -- resumes exactly where it left off, across tab switches, pushed
/// screens, and process death.
///
/// Slices are stored as JSON strings keyed by the RN-side slice key; malformed
/// slices are dropped on read (never propagated to a fresh surface).
enum HostStateStore {
    private static let defaultsKey = "OtaHostSavedState"
    private static let maxSliceBytes = 16 * 1024

    /// Persist one slice (already-serialized JSON object string).
    ///
    /// NOTE: read-modify-write of the whole slice dictionary. Safe today
    /// because brownfield `onMessage` delivery is serialized and only one RN
    /// surface checkpoints at a time; if a second CONCURRENT writer ever
    /// appears, wrap this in a serial queue or a sibling slice can be lost
    /// (Android's per-key SharedPreferences write does not have this hazard).
    static func write(sliceKey: String, stateJson: String) {
        // Size cap mirroring the JS-side guard (host-state.ts); the
        // secret-name denylist lives on the JS side.
        guard stateJson.utf8.count <= maxSliceBytes else {
            print("[HostStateStore] Dropping oversized saved-state slice \(sliceKey)")
            return
        }
        var slices = UserDefaults.standard.dictionary(forKey: defaultsKey) as? [String: String] ?? [:]
        slices[sliceKey] = stateJson
        UserDefaults.standard.set(slices, forKey: defaultsKey)
    }

    /// The whole store as a JSON object of slices: `{"spinner": {...}, ...}`.
    static func readAllJson() -> String {
        let slices = UserDefaults.standard.dictionary(forKey: defaultsKey) as? [String: String] ?? [:]
        var all: [String: Any] = [:]
        for (key, json) in slices {
            guard let data = json.data(using: .utf8),
                  let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { continue }
            all[key] = parsed
        }
        guard let data = try? JSONSerialization.data(withJSONObject: all),
              let string = String(data: data, encoding: .utf8)
        else { return "{}" }
        return string
    }
}
